server/src/main/java/org/opensearch/extensions/ExtensionsManager.java

```diff
@@ -351,16 +315,54 @@ private void extensionsDiscovery() throws IOException {
+     * Loads a single extension
+     * @param extension The extension to be loaded
+     */
+    private void loadExtension(Extension extension) throws IOException {
+        if (extensionIdMap.containsKey(extension.getUniqueId())) {
+            logger.info("Duplicate uniqueId " + extension.getUniqueId() + ". Did not load extension: " + extension);
+        } else {
+            try {
+                DiscoveryExtensionNode discoveryExtension = new DiscoveryExtensionNode(
+                    extension.getName(),
+                    extension.getUniqueId(),
+                    // placeholder for ephemeral id, will change with POC discovery
+                    extension.getUniqueId(),
+                    extension.getHostName(),
+                    extension.getHostAddress(),
+                    new TransportAddress(InetAddress.getByName(extension.getHostAddress()), Integer.parseInt(extension.getPort())),
+                    new HashMap<String, String>(),
+                    Version.fromString(extension.getOpensearchVersion()),
+                    new PluginInfo(
+                        extension.getName(),
+                        extension.getDescription(),
+                        extension.getVersion(),
+                        Version.fromString(extension.getOpensearchVersion()),
+                        extension.getJavaVersion(),
+                        extension.getClassName(),
+                        new ArrayList<String>(),
+                        Boolean.parseBoolean(extension.hasNativeController())
+                    )
+                );
+                extensionIdMap.put(extension.getUniqueId(), discoveryExtension);
+                logger.info("Loaded extension with uniqueId " + extension.getUniqueId() + ": " + extension);
+            } catch (IllegalArgumentException e) {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Iterate through all extensions and initialize them.  Initialized extensions will be added to the {@link #extensions}, and the {@link #namedWriteableRegistry} will be initialized.
      */
-    public void extensionsInitialize() {
-        for (DiscoveryExtension extension : extensionIdMap.values()) {
-            extensionInitialize(extension);
+    public void initialize() {
+        for (DiscoveryExtensionNode extension : extensionIdMap.values()) {
+            initializeExtension(extension);
         }
-        this.namedWriteableRegistry = new ExtensionNamedWriteableRegistry(extensionsInitializedList, transportService);
+        this.namedWriteableRegistry = new ExtensionNamedWriteableRegistry(extensions, transportService);
     }
 
-    private void extensionInitialize(DiscoveryExtension extension) {
+    private void initializeExtension(DiscoveryExtensionNode extension) {
         final CountDownLatch inProgressLatch = new CountDownLatch(1);
```

Comment:
Sorry, I missed that previously, could we replace lath with `CompletableFuture` instead?